### PR TITLE
Cache suspensions

### PR DESCRIPTION
**Pull request by kelunik**
Not sure whether this is useful. If we go that route, we should rename `createSuspension` into `getSuspension`.

### CODE REVIEW

**Review by trowski — APPROVED** [review]
`createSuspension` I think is a fine name. Decorating drivers may still return a new object implementing `Suspension` to implement some fiber-locals, etc.

**Comment by kelunik** [comment]
`create` sounds like it always returns a new object, which it doesn't do anymore then.

**Review by WyriHaximus — APPROVED** [review]

**Comment by trowski** [comment]
> create sounds like it always returns a new object, which it doesn't do anymore then.

Yes, I suppose you're right. My initial thinking was `create` would better indicate the method should be called rather than reusing a prior `Suspension` object by the *consumer* (which may have been associated with a different fiber). However, create certainly doesn't make this obvious and will need to be documented anyway. `get` likely does make more sense then. I'm fine if you want to change that now as part of this PR.

**Review by azjezz — APPROVED** [review]

**Comment by azjezz** [comment]
`getX` also gives the assumption that it always returns the same suspension.

`DriverInterface::suspension(): SuspensionInterface - return, or create suspension for the current fiber`

**Comment by kelunik** [comment]
It _does_ return the same instance, per fiber. It's similar to a `ThreadLocal` in Java.

**Comment by azjezz** [comment]
yea, but I'm talking generally ( as between different fibers ), maybe `DriverInterface::currentSuspension()`?

**Comment by kelunik** [comment]
@trowski @WyriHaximus Renamed and added a documentation comment that the driver must always return the same instance for each fiber.

**Review by trowski — APPROVED** [review]

**Comment by trowski** [comment]
+1 for keeping the old method and deprecating it during 0.2.x.

**Comment by kelunik** [comment]
I've added `EventLoop::createSuspension` again to allow for a smoother transition from 0.1 to 0.2, so libraries can mostly require `^0.1 || ^0.2` instead of directly jumping to `^0.2`.

File: src/EventLoop.php [hunk 1]
```diff
@@ -350,13 +351,31 @@ public static function getDriver(): Driver
...
+     * Returns an object used to suspend and resume execution of the current fiber or {main}.
+     *
+     * Calls from the same fiber will return the same suspension object.
      *
      * @return Suspension
+     *
+     * @deprecated This old name is only kept temporarily to allow smooth transitions from 0.1 to 0.2 and will be
+     *     removed at a later point.
      */
+    #[Deprecated]
```
Comment (azjezz) [inline src/EventLoop.php]: maybe not? 😅 

This doesn't add any benefits over using `@deprecated` tag, PhpStorm can already spot it, and so all other SA tools.

File: composer.json [hunk 2]
```diff
@@ -36,7 +36,8 @@
         "ext-json": "*",
         "phpunit/phpunit": "^9",
         "jetbrains/phpstorm-stubs": "^2019.3",
-        "psalm/phar": "^4.7"
+        "psalm/phar": "^4.7",
+        "jetbrains/phpstorm-attributes": "^1.0"
```
Comment (azjezz) [inline composer.json]: same ^

**Review by azjezz — CHANGES_REQUESTED** [review]

File: src/EventLoop.php (hunk 1 above)
Comment (kelunik) [reply src/EventLoop.php]: Don't really care, I'm fine with either.

Comment (trowski) [reply src/EventLoop.php]: Let's use `@deprecated` to avoid the dependency then, even if it is just dev.

**Review by azjezz — APPROVED** [review]

**Review by WyriHaximus — APPROVED** [review]

**Pull request merged**